packages/data-models/src/definition.js

```diff
@@ -30,11 +98,11 @@ class Definition {
   convertToJSONObject () {
     // eslint-disable-next-line prefer-const
     let result = {
-      text: this.text,
-      language: this.language,
+      text: this._text,
+      languageCode: this._language.toCode(),
       format: this.format,
       lemmaText: this.lemmaText,
-      ID: this.ID
+      ID: this._ID
```

Comment:
I don't really understand the sense of dividing properties to public and private on this stage.
I believe (it is only my thoughts), that one of the reasons to divide propetries to public and private, that
- public could be needed directly in other classes
- private could not be used directly

But you create methods with making `_`like properties public and they could be used directly anyway with additional lines of code.

Another reason to create private properties - is to scope changes of such properties inside one method and don't allow to change it directly. But you change such properties directly (like `_ID`).

Is this additional code is really useful and give us any benefits?

Reply:
Even though javascript doesn't enforce public and private class members, I do think it's helpful to have code standards for identifying the intent because it makes it easy to see when it is being accessed from outside the intended scope. I think using the _ for this is pretty standard and well understood.

Reply:
Then if we decide to use this as a code standard for all outr team - let's define - what class properties should be defined as private and what should be defined as public. 

What are exactly conditions to define a class propety as private? I suggest to define them and print somewhere :)

Reply:
With content-based IDs most of our objects that have content-based IDs should be the value objects (in terms of Domain Driven Design): they should represent a value, or a set of values, and if the value(s) is(are) changed, we should create another instance of the object with new values rather than mutate values within the current object.

This code follows this approach, I believe. The values that affect the ID, and the ID values itself should be immutable. I think the best way to achieve this would be to make the corresponding props private, accessible only from within the class, (well, it does not provide any protection now but the support of "real" private fields is already in Chrome, Edge, and FF with the flag) and provide getters only. This would make the values read only: no one outside the class will be able to mutate their values. So that was the idea.

@balmas, @irina060981: what do you think? Does it make sense?

Reply:
> Then if we decide to use this as a code standard for all out team - let's define - what class properties should be defined as private and what should be defined as public.
> What are exactly conditions to define a class property as private? I suggest to define them and print somewhere :)

On my opinion, it's simple: if the prop is referred to (read or written) outside of the methods of the class, it should be public. If the prop is used by methods of the class only, it should be private. It is very useful during refactoring, for example: we know that private props are used within the class file only, so if we decide to rename it, it would be enough to rename all references within the class. We should be more careful, though, with public members. A public declaration is an indication that this prop is exposed to the outside world and can be used in many places, even outside our own code. So we should change them rarely and be extremely careful while doing so.

Major browsers already start to support the "real" private fields and methods: https://caniuse.com/?search=private. Safari lags behind, unfortunately.

@balmas, @irina060981, what do you think? Does it make sense to you?

Reply:
I agree we should have some guidelines about how and when to use public vs private properties as well as public vs private methods.

Generally, I think we should avoid public properties where possible, instead using getters and setters.

Reply:
>On my opinion, it's simple: if the prop is referred to (read or written) outside of the methods of the class, it should be public. If the prop is used by methods of the class only, it should be private. It is very useful during refactoring, for example: we know that private props are used within the class file only, so if we decide to rename it, it would be enough to rename all references within the class. We should be more careful, though, with public members. A public declaration is an indication that this prop is exposed to the outside world and can be used in many places, even outside our own code. So we should change them rarely and be extremely careful while doing so.

>Generally, I think we should avoid public properties where possible, instead using getters and setters.

I think that at first we should align our understanding what is public/protected and private in JS context.
According to this - https://javascript.info/private-protected-properties-methods

**public properties/methods** 
- are freely accessable, could be read and updated directly by its name outside the class, 
- they could have have getter/setter, but it is not mandatory
- easily used by child classes via inheritance

**protected properties/methods**
- defined with _ (underscore)
- should not be accessed directly by outside
- getter/setter are mandotary
- could make them readonly
- could be inherited by child classes

**private properties/methods**
- defined with # (hash sign)
- could be accessed only inside the class it is defined
- it could not be inherited in child classes, should use getter/setter inside inheritance

According to the article **protected** is only a programmer convension for now, and **private** is a recent addition to the language. We should have a polyfill for controlling it.
for example like this
https://babeljs.io/docs/en/babel-plugin-proposal-private-methods


So according to all this information and your (@kirlat  and @balmas) proposals I believe we shopuld define the following:

- properties could not be `public` at all
- methods could be `public`

- properties could be `protected` (`_`), should have getter/setter, could be used for inheritance, could be used through methods outside the class
- `protected` properties could be readonly (only getter)
- properties could be `private` (`#`) - should have getter/setter only for inherited classes, and no outside access

- methods could be `protected` - to be used inside the class and in children classes (not outside)
- methods could be `private` - to be used only inside the class (not outside)

- we could define class constants using getters

- we should some polyfil for supporting private

I suggest to define some plan/steps of refactoring the code to this standard.
@balmas and @kirlat , what do you think?

Reply:
That's very interesting (thanks for gathering the info!), and I need to think about that a little. I can only say so far that, on my opinion, it's better to avoid Babel for our build process. It will make our build process even more complicated as it is now and I think we should strive to make it simpler instead. It will also increase the build time. It will introduce several dependencies as well. I'm not sure that the benefits it will bring will outweigh that.

Reply:
@kirlat and @irina060981 I think what we decided in our check-in today was the following:

Our general development guidelines should be:

* all properties should be "protected/private", accessed by getter/setter methods
* Until all browsers support # private properties we will not distinguish between protected and private in our code
* We will use _ to name private/protected methods and properties
* When full browser support for # private properties/methods is in place, we should reexamine our code to decide which protected/private methods and properties should become explicitly private

Did I miss anything?

Reply:
I think that's all. I will add that to the documentation.

Reply:
> all properties should be "protected/private", accessed by getter/setter methods

I would maybe make an exception from this rule for the pure JSON-like object that contain data only. They often are read only. We might use these objects to represent data passed or received via the GraphQL API, or to store a set of constants. 

Those objects have no methods (thus no logic of their own), and are used for data storage only. I feel that creating a getter and/or setter for each prop of those might be an overkill, especially for objects in GraphQL that could have a lot of fields. What do you think about that? We can use a special naming convention for such object to distinguish them from the full-fledged OOP class instances.

Reply:
yes that seems sensible to me

Reply:
agree